pyanaconda/modules/network/nm_client.py

```diff
@@ -72,7 +74,74 @@ def nm_client_in_thread():
     finally:
         mainctx.pop_thread_default()
 
-    yield nm_client
+    # Pass a weak reference to the context so that it does not refer it
+    # when we try to destroy it on context exit.
+    nm_client_ref = weakref.ref(nm_client)
+    yield nm_client_ref
+
+    if nm_client:
+        # Pass a reference via a list so it can be removed inside the function.
+        nm_client_holder = [nm_client]
+        del nm_client
+        destroy_nm_client(nm_client_holder)
+
+
+def destroy_nm_client(nm_client_holder):
+    """Destroy NMClient instance.
+
+    NMClient instance will be destroyed if the only refence to it is the one inside
+    the nm_client_holder.
+    Shutting down NMClient properly is tricky.
+    For more information see NetworkManager example examples/python/gi/gmaincontext.py
+    FIXME: use NMClient.wait_shutdown_finish when available
+
+    :param nm_client_holder: list with single NM.NMClient instance
+    :type nm_client_holder: [NM.NMClient]
+    """
+    (nm_client,) = nm_client_holder
+    nm_client_holder.clear()
+    log.debug("[destroy_nm_client]: destroying NMClient %s: pyref=%s, ref_count=%s",
+              nm_client, sys.getrefcount(nm_client), nm_client.ref_count)
+
+    ctx = nm_client.get_main_context()
+
+    finished = []
+
+    def _weak_ref_cb():
+        log.debug("[destroy_nm_client]: context busy watcher is gone")
+        finished.clear()
+        finished.append(True)
+
+    # We take a weak ref on the context-busy-watcher object and give up
+    # our reference on nm_client. This must be the last reference, which initiates
+    # the shutdown of the NMClient.
+    weak_ref = nm_client.get_context_busy_watcher().weak_ref(_weak_ref_cb)
+    del nm_client
+
+    def _timeout_cb(unused):
+        if not finished:
+            # Somebody else holds a reference to the NMClient and keeps
+            # it alive. We cannot properly clean up.
+            log.debug("[destroy_nm_client]: ERROR: timeout waiting for context "
+                      "busy watcher to be gone")
```

Comment:
I think this is a bit confusing - maybe "busy watcher to be/needs to be removed" ? Provided I understand this correctly. :)

Reply:
Actually the message is:
"timeout waiting for context busy watcher to be gone" meaning it is waiting for "context busy watcher". Unfortunately I split the message based on length, not on the sense, which made it confusing for review I guess.